c/meterpreter/source/extensions/winpmem/winpmem.cpp

```diff
@@ -0,0 +1,1077 @@
+/*
+  Copyright 2012-2014 Michael Cohen <scudette@gmail.com>
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+/********************************************************************
+   This is a single binary memory imager for Windows.
+
+   Supported systems:
+    - Windows XPSP2 to Windows 8 inclusive, both 32 bit and 64 bit.
+
+*********************************************************************/
+#include "winpmem.h"
+#include <time.h>
+
+
+__int64 WinPmem::pad(__int64 length) {
+  __int64 count = 1;
+  __int64 start = 0;
+
+  ZeroMemory(buffer_, buffer_size_);
+
+  while(start < length) {
+    DWORD to_write = (DWORD)min(buffer_size_, length - start);
+    DWORD bytes_written;
+
+    if(!WriteFile(out_fd_, buffer_,
+                  to_write, &bytes_written, NULL) ||
+       bytes_written != to_write) {
+      LogLastError(TEXT("Failed to write padding"));
+      goto error;
+    };
+
+    out_offset += bytes_written;
+
+    start += bytes_written;
+    Log(TEXT("."));
+
+    if(!(count % 60)) {
+      Log(TEXT("\n0x%08llX "), start);
+    }
+
+    count ++;
+
+  };
+
+  return 1;
+
+ error:
+  return 0;
+};
+
+__int64 WinPmem::copy_memory(unsigned __int64 start, unsigned __int64 end) {
+  LARGE_INTEGER large_start;
+  __int64 count = 0;
+
+  if (start > max_physical_memory_) {
+    return 0;
+  };
+
+  // Clamp the region to the top of physical memory.
+  if (end > max_physical_memory_) {
+    end = max_physical_memory_;
+  };
+
+  while(start < end) {
+    DWORD to_write = (DWORD)min(buffer_size_, end - start);
+    DWORD bytes_read = 0;
+    DWORD bytes_written = 0;
+
+    large_start.QuadPart = start;
+
+    if(0xFFFFFFFF == SetFilePointerEx(
+       fd_, large_start, NULL, FILE_BEGIN)) {
+      LogError(TEXT("Failed to seek in the pmem device.\n"));
```

Comment:
This file came from the original Winpmem project, and wasn't modified.
I used classes derivation in winpmem_meterpreter.cpp to adjust this code to my needs.
It was done to easily update the code in case of winpmem updates (maybe a bad idea....)